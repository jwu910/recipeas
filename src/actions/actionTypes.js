export const ADD_RECIPE =  'ADD_RECIPE';
export const EDIT_RECIPE = 'EDIT_RECIPE';
export const REMOVE_RECIPE =  'REMOVE_RECIPE';

export const ACTIVATE_EDIT_MODE = 'ACTIVATE_EDIT_MODE';

export const MODIFY_TEMP_TITLE = 'MODIFY_TEMP_INGREDIENT';
export const LOAD_STORED_RECIPE = 'LOAD_STORED_RECIPE';
export const ADD_TEMP_INGREDIENT = 'ADD_TEMP_INGREDIENT';
export const EDIT_TEMP_INGREDIENT = 'EDIT_TEMP_INGREDIENT';
export const REMOVE_TEMP_INGREDIENT = 'REMOVE_TEMP_INGREDIENT';
export const CLEAR_TEMP_RECIPE = 'CLEAR_TEMP_RECIPE';

export const ADD_SEARCH_INGREDIENT_RESULTS = 'ADD_SEARCH_INGREDIENT_RESULTS';
export const CLEAR_SEARCH_INGREDIENT_RESULTS = 'CLEAR_SEARCH_INGREDIENT_RESULTS';
