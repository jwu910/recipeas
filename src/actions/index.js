export const ADD_RECIPE =  'ADD_RECIPE';
export const EDIT_RECIPE = 'EDIT_RECIPE';
export const REMOVE_RECIPE =  'REMOVE_RECIPE';

export const addRecipe = recipe => {
  return {
    type: ADD_RECIPE,
    recipe: recipe
  }
}

export const editRecipe = (targetTitle, recipe) => {
  return {
    type: EDIT_RECIPE,
    targetTitle: targetTitle,
    recipe: recipe
  }
}

export const removeRecipe = recipeTitle => {
  return {
    type: REMOVE_RECIPE,
    recipeTitle: recipeTitle
  }
}
