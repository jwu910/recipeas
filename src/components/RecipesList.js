import React, { Component } from 'react';
import Recipe from '../components/Recipe';
import PlusButton from '../components/PlusButton';
import { Link } from 'react-router-dom';
import { ListGroup, ListGroupItem } from 'react-bootstrap';
import PropTypes from 'prop-types';

class RecipesList extends Component {
  constructor(props) {
    super(props);

    // Makes a list of Recipe components
    this.renderList = this.renderList.bind(this);
  }
  renderList() {
    const {
      removeRecipe,
      loadStoredRecipe,
      activateEditMode
    } = this.props;

    return this.props.recipes.map(recipe => {
      const ingredientsList = recipe.ingredients.map(ingredient => {
        return (
          <ListGroupItem key={ingredient.id}>
            { ingredient.name }
          </ListGroupItem>
        );
      });

      return (
        <ListGroupItem key={recipe.id}>
          <Recipe
            title={recipe.titleDetails.title}
            ingredientsList={ingredientsList}
            removeRecipe={removeRecipe}
            id={recipe.id}
            loadStoredRecipe={() => loadStoredRecipe(recipe)}
            activateEditMode={activateEditMode}
          />
        </ListGroupItem>
      );
    });
  }
  render() {
    return (
      <div>
        <span>
          <h1>
            Recipes
            <Link to="/modify-title">
              <PlusButton className="pull-right"/>
            </Link>
          </h1>
        </span>

        <ListGroup>
          { this.props.recipes ? this.renderList() : 'Add a recipe!' }
        </ListGroup>
      </div>
    );
  }
}

RecipesList.PropTypes = {
  recipes: PropTypes.array.isRequired,
  removeRecipe: PropTypes.func.isRequired
};

export default RecipesList;
