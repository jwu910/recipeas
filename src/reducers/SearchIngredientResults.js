import {
  ADD_SEARCH_INGREDIENT_RESULTS,
  CLEAR_SEARCH_INGREDIENT_RESULTS
} from '../actions/actionTypes';

const searchIngredientResults = (state = {}, action) => {
  switch (action.type) {
    case ADD_SEARCH_INGREDIENT_RESULTS:
      return Object.assign({}, state, action.results);
    case CLEAR_SEARCH_INGREDIENT_RESULTS:
      return {};
    default:
      return state;
  }
};

export default searchIngredientResults;
