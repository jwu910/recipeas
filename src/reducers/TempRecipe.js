import {
  MODIFY_TEMP_TITLE,
  ADD_TEMP_INGREDIENT,
  EDIT_TEMP_INGREDIENT,
  REMOVE_TEMP_INGREDIENT,
  CLEAR_TEMP_RECIPE
} from '../actions';

const initialTempRecipe = {
  titleDetails: {
    title: '',
    allergens: '',
    servings: ''
  },
  ingredients: []
};

const TempRecipe = (state = initialTempRecipe, action) => {
  switch(action.type) {
    // works for adding or editing
    case MODIFY_TEMP_TITLE:
      return Object.assign({}, state, action.titleDetails);
    case ADD_TEMP_INGREDIENT:
      return Object.assign({}, state, {
        ingredients: [
          ...state.ingredients,
          action.ingredient
        ]
      });
    case EDIT_TEMP_INGREDIENT:
      return Object.assign({}, state, {
        ingredients: state.ingredients.map(ingredient => {
          return ingredient.id === action.id ?
            { id: ingredient.id, ...action.ingredient }
            : ingredient;
        })
      });
    case REMOVE_TEMP_INGREDIENT:
      return Object.assign({}, state, {
        ingredients: state.ingredients.filter(ingredient => ingredient.id !== action.id)
      });
    case CLEAR_TEMP_RECIPE:
      return {};
    default:
      return state;
  }
};

export default TempRecipe;
