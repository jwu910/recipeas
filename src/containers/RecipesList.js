import React, { Component } from 'react';
import { addRecipe, editRecipe, removeRecipe } from '../actions/index';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import Recipe from '../components/Recipe';
import RecipeModal from '../components/RecipeModal';
import {
  Button,
  ListGroup,
  ListGroupItem,
  Collapse
} from 'react-bootstrap';

export class RecipesList extends Component {
  constructor(props) {
    super(props);

    this.state = {
      displayAddRecipe: false,
      displayEditRecipe: false,
      activeRecipe: {
        title: '',
        ingredients: ''
      },
      targetTitle: ''
    }
    // methods that affect modal state
    this.handleTitleChange = this.handleTitleChange.bind(this);
    this.handleIngredientsChange = this.handleIngredientsChange.bind(this);
    this.handleSubmit = this.handleSubmit.bind(this);
    this.resetState = this.resetState.bind(this);
    this.fillEditRecipeModal = this.fillEditRecipeModal.bind(this);

    //methods that have to do with modal visibility
    this.showAddRecipe = this.showAddRecipe.bind(this);
    this.hideAddRecipe = this.hideAddRecipe.bind(this);
    this.showEditRecipe = this.showEditRecipe.bind(this);
    this.hideEditRecipe = this.hideEditRecipe.bind(this);

    // Makes a list of Recipe components
    this.renderList = this.renderList.bind(this);


  }
  handleTitleChange(event) {
    this.setState({
      ...this.state,
      activeRecipe: {
        ...this.state.activeRecipe,
        title: event.target.value
      }
    });
  }
  handleIngredientsChange(event) {
    this.setState({
      ...this.state,
      activeRecipe: {
        ...this.state.activeRecipe,
        ingredients: event.target.value
      }
    });
  }
  resetState() {
    this.setState({
      ...this.state,
      activeRecipe: {
        title: '',
        ingredients: ''
      }
    });
  }
  showAddRecipe() {
    this.setState({ ...this.state, displayAddRecipe: true });
  }
  hideAddRecipe() {
    this.setState({ ...this.state, displayAddRecipe: false }, () => this.resetState());
  }
  showEditRecipe() {
    this.setState({ ...this.state, displayEditRecipe: true });
  }
  hideEditRecipe() {
    this.setState({ ...this.state, displayEditRecipe: false }, () => this.resetState());
  }
  fillEditRecipeModal(title) {
    const targetRecipe = this.props.recipes.find(recipe => {
      return recipe.title === title;
    });

    this.setState({
      ...this.state,
      activeRecipe: {
        title: targetRecipe.title,
        ingredients: targetRecipe.ingredients
      },
      targetTitle: targetRecipe.title
    }, () => this.showEditRecipe());
  }
  handleSubmit(event) {
    event.preventDefault();

    if (this.state.displayAddRecipe) {
      this.props.addRecipe(this.state.activeRecipe);
    }
    else if (this.state.displayEditRecipe) {
      this.props.editRecipe(this.state.targetTitle, this.state.activeRecipe);
    }
  }
  renderList() {
    const { removeRecipe } = this.props;

    return this.props.recipes.map(recipe => {
      const ingredientsList = recipe.ingredients.split(',').map(ingredient => {
        return (
          <ListGroupItem key={recipe.title + ingredient}>
            { ingredient }
          </ListGroupItem>
        )
      });

      return (
        <ListGroupItem key={recipe.title}>
          <Collapse in={recipe.open}>
            <Recipe
              title={recipe.title}
              ingredients={ingredientsList}
              showEditRecipe={this.showEditRecipe}
              removeRecipe={removeRecipe}
              fillEditRecipeModal={this.fillEditRecipeModal}
            />
        </Collapse>
        </ListGroupItem>
      );
    });
  }
  render() {
    return (
      <div className="container">
        <h1>Recipes</h1>
        <ListGroup>
          { this.renderList() }
        </ListGroup>
        <Button bsStyle="primary" onClick={this.showAddRecipe}>Add Recipe</Button>
         <RecipeModal
           modalTitle="Add Recipe"
           show={this.state.displayAddRecipe}
           onHide={this.hideAddRecipe}
           recipeTitle={this.state.activeRecipe.title}
           recipeIngredients={this.state.activeRecipe.ingredients}
           handleTitleChange={this.handleTitleChange}
           handleIngredientsChange={this.handleIngredientsChange}
           handleSubmit={this.handleSubmit}
         />
         <RecipeModal
           modalTitle="Edit Recipe"
           show={this.state.displayEditRecipe}
           onHide={this.hideEditRecipe}
           recipeTitle={this.state.activeRecipe.title}
           recipeIngredients={this.state.activeRecipe.ingredients}
           handleTitleChange={this.handleTitleChange}
           handleIngredientsChange={this.handleIngredientsChange}
           handleSubmit={this.handleSubmit}
         />
      </div>
    );
  }
}

const mapStateToProps = state => {
  return {
    recipes: state.recipes
  };
}

const mapDispatchToProps = dispatch => {
  return bindActionCreators({ addRecipe, editRecipe, removeRecipe }, dispatch);

}
export default connect(
  mapStateToProps,
  mapDispatchToProps
)(RecipesList);
