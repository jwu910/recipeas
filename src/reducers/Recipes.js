import { ADD_RECIPE, EDIT_RECIPE, REMOVE_RECIPE } from '../actions';

export const defaultRecipes = [
  {
    title: 'Pumpkin Pie',
    ingredients: 'Pumpkin Puree, Sweetened Condensed Milk, Eggs, Pumpkin Pie Spice, Pie Crust'
  },
  {
    title: 'Spaghetti',
    ingredients: 'Noodles, Pasta Sauce, Meatballs'
  },
  {
    title: 'Peanut Butter Mug Cake',
    ingredients: 'Peanut Butter, Eggs, Baking Powder, Sugar'
  }
];

const Recipes = (state = defaultRecipes, action) => {
  switch (action.type) {
    case ADD_RECIPE:
      return [
        ...state,
        action.recipe
      ]
    case EDIT_RECIPE:
      return state.map(recipe => {
        return recipe.title === action.targetTitle ?
          action.recipe
          : recipe;
      });
    case REMOVE_RECIPE:
      return state.filter(recipe => recipe.title !== action.recipeTitle);
    default:
      return state;
  }
}

export default Recipes;
