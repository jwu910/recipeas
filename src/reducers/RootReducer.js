import { combineReducers } from 'redux';
import Recipes from './Recipes';
import TempRecipe from './TempRecipe';
import SearchIngredientResults from './SearchIngredientResults';

const RootReducer = combineReducers({
  recipes: Recipes,
  tempRecipe: TempRecipe,
  results: SearchIngredientResults
});

export default RootReducer;
