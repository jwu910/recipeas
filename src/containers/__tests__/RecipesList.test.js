import React from 'react';
import renderer from 'react-test-renderer';
import { shallow, mount } from 'enzyme';
import { RecipesList } from '../RecipesList';
import { defaultRecipes } from '../../reducers/Recipes';
import { removeRecipe } from '../../actions/index';
import configureStore from 'redux-mock-store';
import { Provider } from 'react-redux';
import { Button } from 'react-bootstrap';

describe('Recipes List', () => {
  test('Renders without crashing', () => {
    const mockStore = configureStore()({
      recipes: defaultRecipes
    });

    const component = renderer.create(
      <Provider store={mockStore}>
        <RecipesList recipes={defaultRecipes} removeRecipe={removeRecipe} />
      </Provider>
    );
    const tree = component.toJSON();

    expect(tree).toMatchSnapshot();
  });

  test('Receives default state from store', () => {
    const mockStore = configureStore()({
      recipes: defaultRecipes
    });

    const wrapper = shallow(
      <Provider store={mockStore}>
        <RecipesList recipes={defaultRecipes} removeRecipe={removeRecipe} />
      </Provider>
    );

    expect(wrapper.props().recipes).toEqual(defaultRecipes);
  });

  test('Add recipe button shows AddRecipe modal', () => {
    const wrapper = shallow(
      <RecipesList recipes={defaultRecipes} removeRecipe={removeRecipe} />
    );
    wrapper.find(Button).simulate('click');

    // showAddRecipe controls visibility of AddRecipe modal
    expect(wrapper.state().displayAddRecipe).toEqual(true);
  });

  test('Edit button shows EditRecipe modal', () => {
    const wrapper = mount(
      <RecipesList recipes={defaultRecipes} removeRecipe={removeRecipe} />
    );
    wrapper.find('.edit-recipe').first().simulate('click');

    expect(wrapper.state().displayEditRecipe).toEqual(true);
  });

  // TODO: maybe a delete button test is necessary?
});
