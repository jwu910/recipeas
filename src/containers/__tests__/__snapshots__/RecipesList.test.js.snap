// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Recipes List Renders without crashing 1`] = `
<div
  className="container"
>
  <h1>
    Recipes
  </h1>
  <ul
    className="list-group"
  >
    <li
      className="list-group-item"
    >
      <div
        className="recipe-container"
      >
        <h4
          className="recipe-title"
          onClick={[Function]}
        >
          Pumpkin Pie
        </h4>
        <div
          aria-expanded={null}
          className="collapse"
        >
           
          <div
            className="well"
          >
            <ul>
              <span
                className="list-group-item"
              >
                Pumpkin Puree
              </span>
              <span
                className="list-group-item"
              >
                 Sweetened Condensed Milk
              </span>
              <span
                className="list-group-item"
              >
                 Eggs
              </span>
              <span
                className="list-group-item"
              >
                 Pumpkin Pie Spice
              </span>
              <span
                className="list-group-item"
              >
                 Pie Crust
              </span>
            </ul>
            <button
              className="edit-recipe btn btn-default"
              disabled={false}
              onClick={[Function]}
              type="button"
            >
              Edit
            </button>
            <button
              className="delete-recipe btn btn-danger"
              disabled={false}
              onClick={[Function]}
              type="button"
            >
              Delete
            </button>
          </div>
        </div>
      </div>
    </li>
    <li
      className="list-group-item"
    >
      <div
        className="recipe-container"
      >
        <h4
          className="recipe-title"
          onClick={[Function]}
        >
          Spaghetti
        </h4>
        <div
          aria-expanded={null}
          className="collapse"
        >
           
          <div
            className="well"
          >
            <ul>
              <span
                className="list-group-item"
              >
                Noodles
              </span>
              <span
                className="list-group-item"
              >
                 Pasta Sauce
              </span>
              <span
                className="list-group-item"
              >
                 Meatballs
              </span>
            </ul>
            <button
              className="edit-recipe btn btn-default"
              disabled={false}
              onClick={[Function]}
              type="button"
            >
              Edit
            </button>
            <button
              className="delete-recipe btn btn-danger"
              disabled={false}
              onClick={[Function]}
              type="button"
            >
              Delete
            </button>
          </div>
        </div>
      </div>
    </li>
    <li
      className="list-group-item"
    >
      <div
        className="recipe-container"
      >
        <h4
          className="recipe-title"
          onClick={[Function]}
        >
          Peanut Butter Mug Cake
        </h4>
        <div
          aria-expanded={null}
          className="collapse"
        >
           
          <div
            className="well"
          >
            <ul>
              <span
                className="list-group-item"
              >
                Peanut Butter
              </span>
              <span
                className="list-group-item"
              >
                 Eggs
              </span>
              <span
                className="list-group-item"
              >
                 Baking Powder
              </span>
              <span
                className="list-group-item"
              >
                 Sugar
              </span>
            </ul>
            <button
              className="edit-recipe btn btn-default"
              disabled={false}
              onClick={[Function]}
              type="button"
            >
              Edit
            </button>
            <button
              className="delete-recipe btn btn-danger"
              disabled={false}
              onClick={[Function]}
              type="button"
            >
              Delete
            </button>
          </div>
        </div>
      </div>
    </li>
  </ul>
  <button
    className="btn btn-primary"
    disabled={false}
    onClick={[Function]}
    type="button"
  >
    Add Recipe
  </button>
</div>
`;
